[Index] Apply RelationContainedBy role to references contained by VarDecl.

ileitch opened the pull request:
References associated with a `VarDecl` had no `RelationContainedBy` role, resulting in "orphaned" references. From the perspective of identifying unused code (in tools using the index, like [Periphery](https://github.com/peripheryapp/periphery)), this made it impossible to identify that a variable's type, initializer and custom attributes are associated with the variable.

Resolves #56163

---

Notes to reviewers:
* This is my first contribution to Swift, extra scrutiny needed :sweat_smile:
* I've added some inline comments to explain my reasoning

ileitch committed:
[Index] Apply RelationContainedBy role to references contained by VarDecl.

References associated with a `VarDecl` had no `RelationContainedBy` role, resulting in "orphaned" references. From the perspective of identifying unused code (in tools using the index, like [Periphery](https://github.com/peripheryapp/periphery)), this made it impossible to identify that a variable's type, initializer and custom attributes are associated with the variable.

Resolves: [SR-13766](https://bugs.swift.org/browse/SR-13766)

ileitch commented on include/swift/IDE/SourceEntityWalker.h:
```diff
@@ -82,11 +82,26 @@ class SourceEntityWalker {
+
+  /// This method is called immediately after \c walkToDeclPost or after \c
+  /// walkToDeclPre if it returns false.
+  virtual void endBalancedDeclVisit(Decl *D){};
```
I don't _love_ the addition of these two new methods, however I needed a place to setup the `ContainerStack` before `SemaAnnotator::walkToDeclPre` begins walking. I deliberately haven't alluded to their intended use in their naming, as the `ContainerStack` is solely a concern of `IndexSwiftASTWalker`. I also needed their calls to be balanced, unlike `walkToDeclPre/Post`. I noticed some existing logic and commits dealing with other call balance issues, so it seems to me that the nature of `ASTWalker` to not always balance `walkToDeclPre/Post` calls is a little problematic. Two possible solutions to this would be to move these new methods down into `ASTWalker`, or modify `ASTWalker` to always call `walkToDeclPost`, but indicate the return value of `walkToDeclPre` with a new argument. I think both of these approaches would be far too invasive for my first ever contribution, and certainly outside the scope of this PR.

benlangmuir replied:
Yeah, I agree this is unfortunate.  I kind of like the idea of changing ASTWalker and SourceEntityWalker to always call the Post function, but it could be a fairly invasive change - I'm not sure how deeply that assumption is baked into the existing subclasses - so I'm not sure one way or the other.

@bnbarham or @rintaro do you have any suggestions or thoughts here?

rintaro replied:
(Without closely looking into the patch)
In general, _not_ calling `walkToDeclPost` if `walkToDeclPre` returned false sounds like the correct behavior to me. It's like calling `button.push()` doesn't/shouldn't invoke a `onPush` event in UI programming. Since the user (i.e. the type implementing `walkToDeclPre`) have the complete control of whether to visit the children, it should be easy to invoke the balanced operation manually.

```cpp
bool walkToDeclPre(Decl *D) {
  callBalancedBegin(D);

  if (condition) {
    callBalancedEnd(D);

    // don't visit children.
    return false;
  }

  // visit children.
  return true;
}

bool walkToDeclPost(Decl *D) {
  callBalancedEnd(D);
}
```

ileitch replied:
Unfortunately I can't push all this down into `IndexSwiftASTWalker` because `SemaAnnotator::walkToDeclPre` calls `handleCustomAttributes` (plus some other stuff) prior to calling `SourceEntityWalker::walkToDeclPre`. I need the container setup before `handleCustomAttributes` is called so that the var also contains property wrapper references.

bnbarham replied:
So first off - I only just noticed that you had updated your forum post, sorry for missing these! I was expecting notifications, but I forgot to set "watching" on the topic. Feel free to @ me/ping on the Swift Slack if I'm unresponsive.

Looks like we actually want the contained by relationship for references in:
  * Declaration type
  * Initializers
  * Custom attributes

I missed that `handleCustomAttributes` was being called after the `beginBalancesDeclVisit` call when first reviewing. The idea of the SourceEntityWalker is to walk in source order, so it's a little weird to have a decl visit method that would be out of source order (eg. the VarDecl and then the custom attribute in this case).

You mention in the thread the possibility of new `walkToContainingDeclPre/Post` methods. I'm thinking something like that is probably the go, along with custom walks for the PDB and pattern in SemaAnnotator (which would call them). Perhaps `visitContainerPre/Post`? It would have to be called multiple times to handle eg.
```
var (a, b): (Int, String) = (foo(), bar())
```
Where the visit order would be `a`, `b`, `Int`, `String`, `foo`, `bar`. Ie. `pre(a) -> Int -> post(a) -> pre(b) -> String -> post(b) -> pre(a) -> foo -> post(a) -> pre(b) -> bar -> post(b)`

@benlangmuir any thoughts?

ileitch replied:
> So first off - I only just noticed that you had updated your forum post, sorry for missing these! I was expecting notifications, but I forgot to set "watching" on the topic. Feel free to @ me/ping on the Swift Slack if I'm unresponsive.

No worries at all! I wasn't aware of there being a public Swift Slack, or is that some internal Apple thing?

> I missed that handleCustomAttributes was being called after the beginBalancesDeclVisit call when first reviewing. The idea of the SourceEntityWalker is to walk in source order, so it's a little weird to have a decl visit method that would be out of source order (eg. the VarDecl and then the custom attribute in this case).

I agree that it's a little weird, though I can't really think of a great way to avoid it. Perhaps `beginBalancedDeclVisit` could be renamed to indicate it's not necessarily in source order?

> You mention in the thread the possibility of new walkToContainingDeclPre/Post methods.

Yes that was my initial plan, however I found that messing with the existing walking got messy very quickly. That's why I went with the `begin/endBalancedDeclVisit` approach, it allows me to just save the container decl on a stack, and leave all the existing walking logic the same.

Note that I deliberately didn't end up going with `walkToContainingDeclPre/Post` as the concept of a container is solely a concern of `IndexSwiftASTWalker`. Implementing a `walkToContainingDeclPre/Post` in `SemaAnnotator` implies that it's only only to be called for container typed, which means I'd need to perform duplicate type casts in both `SemaAnnotator` and `IndexSwiftASTWalker`.

On the destructuring front, I've actually got something working that doesn't involve a custom walk. In a nutshell, it identifies important "singposts" (Expr, TypeRepr) in the pattern binding, along with the `VarDecl` they're associated with. Then, in `walkToExprPre/walkToTypeReprPre` it checks if the `Expr/TypeRepr` is a known "signpost", and if so, it activates the associated `VarDecl`. Then when the normal walk comes to request the container, it just returns whatever the current active `VarDecl` is. I should have it ready for review by tomorrow.

bnbarham replied:
> No worries at all! I wasn't aware of there being a public Swift Slack, or is that some internal Apple thing?

Ah yep, sorry about that! @/DM on Swift Forums would be ideal :)

ileitch commented on lib/Index/Index.cpp:
```diff
@@ -440,6 +451,43 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+  /// Currently both \c a and \c b will contain references to \c Int, \c String,
+  /// \c x and \c y. Future work could improve upon this to try match up the
+  /// indices, such that \c a would only contain references to \c Int and \c x,
+  /// and \c b references to \c String and \c y.
```
As noted here, I would like to address this in the future. However, after some initial investigation my gut tells me that this might be quite tricky to implement, and even if possible, perhaps too brittle to warrant the complexity. If anyone can think of a nice way to achieve this, I'd love to hear it.

benlangmuir replied:
I'm concerned about this multiplicative factor.  Your tests show the quadratic behaviour in the number of decls:
```swift
let (tupleElementA, tupleElementB): (Int, String) = (intValue, stringValue)
```

We are going to get `O(n^2)` relations from this, where `n` is the number of decls.

But there is also the number of references in any of those initializer expressions:
```swift
let ... = (intValue + intValue + ...m... + intValue, ...)
```

So if `m` is the number of references in the longest initializer expression for any of those decls, we end up with `O(n^2 + n*m)` total.

ileitch replied:
Yes that's definitely concerning, though my reasoning was that the vast majority of cases are going to have only a few tuple elements, and that the extra references wouldn't be a cause for concern. However in hindsight that's probably not a defensible argument. I'll take another look into applying the references to the correct elements. If that's not possible however, I can think of a few other solutions:

1. Don't do anything for tuples, keep the current behavior.
2. Use on the first element as the container 🥴 
3. Emit the `PatternBinding` into the index as an implicit definition to contain the declarations and references? I'm not sure if this would even be possible.

Can you think of any other solutions?

benlangmuir replied:
I'd prefer not to do (2), which seems actively wrong.

(3) seems workable to me, although not as nice as actually getting the accurate VarDecl parent. Something like

```
PatternBindingDecl
VarDecl childOf PBD
Type containedBy PBD
"Reference from initializer expression" containedBy PBD
```

On the plus side, this is straightforward to implement (I think), and the distinction between containedBy PBD and containedBy VarDecl seems uninteresting most of the time - actually very few queries even use containedBy relations in my experience.  On the minus side, it's less precise about which vardecl things are attached to, and it introduces a weird declaration kind into the index. Basically it forces the client code to worry about this weird complication.

bnbarham replied:
Ah, I didn't realise you wanted the initializer to be contained by the VarDecl as well (though it makes sense for your use case in hindsight). I'd prefer (3) over adding the references to all decls in the PDB.

I don't think linking up to accurate VarDecls would be *too* bad. You could implement a custom walker for the PDB and Pattern in `IndexASTWalker` (see `Traversal::visitPatternBindingDecl` and `Traversal::visitTypedPattern`). Though it would be a little unfortunate that we would then have two similar-but-not-quite walks.

Another question is what we would want to do for something like `var (a, b) = foo()` though - should `foo` be contained by both `a` and `b` in that case? Or maybe we should add the PDB regardless and have `foo` contained by PDB:

```
var (a, b) = foo()
var (c, d) = (bar(), baz())
```
->
```
PDB1 for (a, b)
a childOf PDB1
b childOf PDB1
foo containedBy PDB

PDB2 for (c, d)
c childOf PDB2
d childOf PDB2
bar containedBy c
baz containedBy d
```

ileitch replied:
After a little thought, I think (3) has the potential to cause a fair amount of breakage for downstream users. At least for Periphery, I'd need to make some changes to support this. I'd likely need to relate each var to every reference anyway, so we've just pushed the quadratic behavior down the line.

There's also the question of what `clang::index::SymbolInfo::Kind` the PBD definition should have. I guess it could either be a `Variable`, or `*Property` to match the `VarDecl`, but then we'd still need some way to communicate that it _isn't really_. The other option is adding a new type, which seems like a no-go.

So perhaps the only viable options are (1) or (4) - matching the references to the decls.

> Another question is what we would want to do for something like var (a, b) = foo() though - should foo be contained by both a and b in that case?

Yes, I believe so. I feels correct to me? I'm also not sure what an alternative would be in this case.

ileitch commented on test/Index/expressions.swift:
```diff
@@ -17,36 +17,6 @@ func test(_ o: P1?) {
   }
 }
 
```
The following is now covered by roles-containing.swift

ileitch commented on test/Index/roles.swift:
```diff
@@ -381,47 +382,6 @@ let _ = aSubInstance.foo()
 let contained = 2
 // CHECK: [[@LINE-1]]:5 | variable/Swift | contained | s:14swift_ide_test9containedSivp | Def | rel: 0
 
```
The following is now covered by roles-containing.swift

ileitch commented:
@bnbarham @benlangmuir

ileitch commented on include/swift/IDE/SourceEntityWalker.h:
```diff
@@ -82,11 +82,26 @@ class SourceEntityWalker {
   /// children.  If it returns false, the subtree is skipped.
-  virtual bool walkToDeclPre(Decl *D, CharSourceRange Range) { return true; }
+  virtual bool walkToDeclPre(Decl *D, CharSourceRange Range) {
+    beginBalancedDeclVisit(D);
```
I added them in the case that a subclass doesn't override `walkToDeclPre/Post`, but does override `begin/endBalancedDeclVisit`. I added them for correctness really, currently they don't serve any practical benefit. I'm happy to remove them if you'd like.

benlangmuir replied:
Won't this cause `beginBalancedDeclVisit` to get called twice if you don't override `walkToDeclPre` since SemaAnnotator will call it directly, then call `walkToDeclPre` which will come here and call it again?

ileitch replied:
Ah yes, that's a good point. I'll remove these calls, and update the documentation for `walkToDeclPre/Post` to mention that it's the job of the subclass to implement calls to `begin/endBalancedDeclVisit`.

ileitch replied:
No, you're quite right. It's me that was confused.

ileitch commented on lib/Index/Index.cpp:
```diff
@@ -440,6 +451,43 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+
+      if (VDs.size() > 0)
+        ContainerStack.push_back({VDs, D});
+    } else if (auto *AFD = dyn_cast<AbstractFunctionDecl>(D)) {
```
I believe the previous check was needed because the `EntitiesStack` could contain reference entities. Whether you'd actually have a `AbstractFunctionDecl` on there that doesn't already have the `Definition` role, I'm not sure. For non-reference entities the `Definition` role is always applied: https://github.com/apple/swift/blob/main/lib/Index/Index.cpp#L1326

In the case of `ContainerStack` entities we don't actually call `initIndexSymbol` as it's not necessary, we don't create a `IndexSymbol` for them. These container declarations will exist both in `ContainerStack`, and in `EntitiesStack`. Those in the `EntitiesStack` _do_ have the associated `IndexSymbol`.

This was the reason I created the `ContainerStack`, as the `EntitiesStack` was already used for a multiple purposes: `getParentDecl` and `getContainingDecl` - which aren't necessarily the same thing. Trying to expand the scope of what could be considered a container declaration whilst still using `EntitiesStack` caused issues. It was just much simpler and cleaner to introduce a new stack solely for containers.

bnbarham commented on test/Index/roles-contained.swift:
```diff
@@ -0,0 +1,142 @@
+// RUN: %target-swift-ide-test -print-indexed-symbols -include-locals -source-filename %s | %FileCheck %s
```
Some more test ideas:
  - TuplePattern with an init that returns a tuple, ie. `let (a, b): (Int, String) = returnsTuple()`
  - Multiple patterns/inits in PDB, ie. `let (a, b) = foo(), (c, d) = bar()`

bnbarham commented on lib/Index/Index.cpp:
```diff
@@ -468,26 +494,127 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
       ContainerStack.pop_back();
   }
 
+  void identifyPatternBindingSignposts(PatternBindingDecl *PBD,
+                                       ContainerStackEntry &Entry) {
```
Is there a reason we couldn't just associate a TypeRepr/Expr directly to VarDecl (if one) or Pattern (if on all) here?

ileitch commented on test/Index/roles-contained.swift:
```diff
@@ -13,16 +15,18 @@ struct Wrapped<T> {
+// CHECK: [[@LINE-2]]:14 | variable/Swift | intValue | {{.*}} | Ref,Read | rel: 0
+
 let typedProperty: Int = 1
-// CHECK: [[@LINE-1]]:5 | variable/Swift | typedProperty | {{.*}} | Def | rel: 0
```
I removed some of these `Def` checks to keep this file focused on `RelCont`. `Def` should already be suitably covered by other tests.

bnbarham commented on include/swift/IDE/SourceEntityWalker.h:
```diff
@@ -88,6 +88,15 @@ class SourceEntityWalker {
   /// false, the remaining traversal is terminated and returns failure.
   virtual bool walkToDeclPost(Decl *D) { return true; }
 
+  /// This method is called immediately before \c walkToDeclPre. It is guranteed
+  /// to be called in balance with its counterpart \c endBalancedDeclVisit even
+  /// if \c walkToDeclPre returns false.
+  virtual void beginBalancedDeclVisit(Decl *D){};
+
+  /// This method is called immediately after \c walkToDeclPost or after \c
+  /// walkToDeclPre if it returns false.
+  virtual void endBalancedDeclVisit(Decl *D){};
```
guranteed  -> guaranteed

@benlangmuir and @rintaro do you have any ideas for naming these? They're not *just* balanced but will have a separate order to `walkTo*` and may be called on more decls than `walkToDecl*`.

Ie. The `TypeRepr`/`Expr`s in custom attributes will be walked before `walkToDecl*` for the attributed decl, but between `*BalancedDeclVisit`.

And another example is if `shouldWalkInactiveConfigRegion` is set then `*BalancedDeclVisit` will be called for an `IfConfigDecl` but `walkToDecl*` will not.

ileitch replied:
`*LookaheadDeclVisit` perhaps?

benlangmuir replied:
> They're not just balanced but will have a separate order to walkTo* and may be called on more decls than walkToDecl*.
> Ie. The TypeRepr/Exprs in custom attributes will be walked before walkToDecl* for the attributed decl, but between *BalancedDeclVisit.
> And another example is if shouldWalkInactiveConfigRegion is set then *BalancedDeclVisit will be called for an IfConfigDecl but walkToDecl* will not.

Interesting, I hadn't noticed this.  There may be good reasons for having a difference in what decls will get the callbacks, but we should be able to document what the difference is.  I do think we need to eliminate the differences in walk _order_ though - that just seems like a foot gun.

ileitch replied:
>  I do think we need to eliminate the differences in walk order though - that just seems like a foot gun.

It seems to me (from a newcomers perspective), that perhaps `IndexSwiftASTWalker` has outgrown the [stated design goals](https://github.com/apple/swift/blob/main/include/swift/IDE/SourceEntityWalker.h#L43-L45) of `SemaAnnotator/SourceEntityWalker`. Due to the introduction of property wrappers, simply walking in source order is no longer practical, at least in order to achieve the index structure this PR introduces. It also seems `IndexSwiftASTWalker` and `SourceEntityWalker` have diverged in regards to implicit declarations, `IndexSwiftASTWalker` does now index some implicit declarations (constructors, accessors). I don't really see an easy solution to this, other than a huge refactor to decouple `IndexSwiftASTWalker` and `SourceEntityWalker`.

ileitch replied:
Another instance of friction is custom attribute initializers. `SemaAnnotator::handleCustomAttributes` [explicitly ignores](https://github.com/apple/swift/blob/main/lib/IDE/SourceEntityWalker.cpp#L644) implicit initializers, whereas `IndexSwiftASTWalker` [explicitly adds them](https://github.com/apple/swift/blob/main/lib/Index/Index.cpp#L343).

bnbarham replied:
> `IndexSwiftASTWalker` does now index some implicit declarations (constructors, accessors)

I believe implicit constructors are walked by `SourceEntityWalker` as well (see `SemaAnnotator::shouldIgnore`).

> Another instance of friction is custom attribute initializers. `SemaAnnotator` explicitly ignores implicit initializers, whereas `IndexSwiftASTWalker` explicitly adds them.

In general I think we probably just want to add an option to `SemaAnnotator` that walks more implicit nodes, rather than customise the behaviour in the subclasses. Though I'm not sure if anything *wouldn't* want to walk these cases.

Regardless, not something to tackle in this PR. Some name that makes it clear that it's visiting in the ASTWalker order versus source order and a clear comment seems reasonable to me. Another alternative could be to implement all the container logic in `SourceEntityWalker`, hide these new methods (private + friend SemaAnnotator), and move `getContainer` and `getParent` to `SourceEntityWalker`.

benlangmuir replied:
> Some name that makes it clear that it's visiting in the ASTWalker order versus source order and a clear comment seems reasonable to me. Another alternative could be to implement all the container logic in SourceEntityWalker, hide these new methods (private + friend SemaAnnotator), and move getContainer and getParent to SourceEntityWalker.

Okay I could accept either of these options, I think.

ileitch replied:
OK, I've improved the documentation for now. Pushing all this down into `SourceEntityWalker` sounds like a good idea though, I'll do that in a follow-up PR.

bnbarham commented on lib/Index/Index.cpp:
```diff
@@ -153,6 +154,20 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
   bool isSystemModule = false;
+
+  struct ContainerStackEntry {
+    Decl *const VisitedDecl;
+    // These vector capacities represent the number of declarations that may
+    // be associated with the key type. In most cases there will be only one; a
+    // function or single named pattern binding. However in the case of an
+    // expanded tuple pattern binding, there will be multiple.
+    SmallVector<Decl *, 6> ActiveDecls{};
+    std::map<Expr *, SmallVector<Decl *, 6>> ExprDecls{};
+    std::map<TypeRepr *, SmallVector<Decl *, 6>> TypeReprDecls{};
+    std::map<Pattern *, SmallVector<Decl *, 6>> PatternDecls{};
```
I suppose there's no harm in marking the pointer const, but I probably wouldn't bother (the vector and maps are still mutable).

Also, given that's there's normally only a single `Decl *`, it seems a bit wasteful to use a `SmallVector<... 6>` as the value in the maps. What do you think about something like:
```
struct ContainerStackEntry {
  Decl *VisitedDecl;
  llvm::PointerUnion<Decl *, TuplePattern *> Active;
  // Don't both storing anything if there's no corresponding decls (ie. all AnyPatterns)
  DenseMap<void *, llvm::PointerUnion<Decl *, TuplePattern *>> DeclMap;
};
```
?

`addContainedByRelationIfContained` could then handle either the `Decl*` or `TuplePattern*`.

And speaking of patterns, do we actually need to keep the map decls for a pattern? The corresponding TypeRepr/Expr will have a mapping and custom attributes can only apply for a single var, so that would already be covered by the active case.

ileitch replied:
> Also, given that's there's normally only a single Decl *, it seems a bit wasteful to use a SmallVector<... 6> as the value in the maps. What do you think about something like:

I agree it's not ideal, though I don't believe your proposed alternative would work. There can be more than one active decl, and they're not always associated with a `TuplePattern`, e.g:
```
let (a, b) = someFunc() // both a and b are active so that each contains someFunc() references.
```
```
func someFunc() {
    let _: Int = 1 // The active container for Int is someFunc(), as the AnyPattern has no VarDecl.
}
```

I did however condense the three maps into one using `void *` (I wasn't sure if using `void *` was considered good c++ or not 😄).

> And speaking of patterns, do we actually need to keep the map decls for a pattern?

When we've a single var in the binding (or an AnyPattern), we just associate the Pattern and decl and ignore the TypeReps and Exprs; whatever types and expressions that follow all need to be contained by the single var anyway. This greatly simplifies [this logic here](https://github.com/apple/swift/pull/38883/commits/49e854dd3a56a02a8f84c0fddfa79f94dfd4ade1#diff-187a51395f73cbba56fa55021f185ae2a88b30e4a7306de23800aba8a235c634R551-R589), which can focus solely on unpacking expanded tuple bindings.

bnbarham replied:
> I agree it's not ideal, though I don't believe your proposed alternative would work. There can be more than one active decl, and they're not always associated with a TuplePattern, e.g:

When is there more than one decl but those aren't all from a TuplePattern? In your example both `(a, b)` are part of the TuplePattern. Then for the AnyPattern case that would just be the `someFunc` Decl (both able to be stored as a `PointerUnion<Decl *, TuplePattern *>`. Ie. my suggestion was essentially that we either associate the single Decl or the TuplePattern.

> When we've a single var in the binding (or an AnyPattern), we just associate the Pattern and decl and ignore the TypeReps and Exprs; whatever types and expressions that follow all need to be contained by the single var anyway

I'm still missing why associating the Pattern is required there - can't we just set the active decl to the single var in that case?

ileitch replied:
I got it working using a `PointerUnion`, please see the latest commit. I think your suggestion of using `TuplePattern` was throwing me off a little, as there's actually other patterns at play here. However we can just use `Pattern` instead, and avoid the need for an extra type in the union:

```
let (c, d) = someFunc() // TuplePattern
let (a, b): (Int, Int) = ... // TypedPattern
```

> I'm still missing why associating the Pattern is required there - can't we just set the active decl to the single var in that case?

Currently when we encounter a `TypedPattern` with a `TupleTypeRepr` we can iterate through the type elements and assume we've a matching number of var decls. However given the binding `let foo: (Int, Int) = (1, 1)` that's not a safe assumption.

So instead of explicitly handling this scenario (along with `AnyPattern` and maybe others) it's simpler and more future proof to just associate the active decl with the `Pattern` when we encounter a single `VarDecl`.

bnbarham replied:
So there are other `Patterns`, but we're handling them by creating the mappings in `identifyPatternBindingContainerDecls`. For the actual mappings themselves we only care about either a single `Decl`, or a `TuplePattern` that contains all the `NamedPatterns`. The `TuplePattern` *has* to contain all NamedPatterns` as otherwise we would have mapped to `Decl` (or the current container).

This also means that `ActiveDecls` can just be `Active` and a `PointerUnion<Decl *, TuplePattern *>` instead. 

To hopefully make that a little more concrete, the AST for `let (a, b): (Int, String) = (expr1, expr2)` will look something like:
```
PDB
  TypedPattern
    TuplePattern
      NamedPattern(a)
      NamedPattern(b)
    TupleTypeRepr
      IdentTypeRepr(Int)
      IdentTypeRepr(String)
  TupleExpr
    expr1
    expr2
```

So in this case we'd want the following mappings:
  - Int/expr1 -> Decl(a)
  - String/expr2 -> Decl(b)

The `TypedPattern` IMO should always map to `nullptr`, which would allow us to use the same stack for `getParent` (ie. a mapping of `nullptr` == use container, which you currently do for the `AnyPattern` case).

Following from this, looking at `let a: (Int, Int) = (1, 1)` the AST would be:
```
PDB
  TypedPattern
    NamedPattern
  TupleTypeRepr
    ...
  TupleExpr
    ...
```

If the sub pattern of the `TypedPattern` is not a `TuplePattern`, then we can just create a mapping:
  - TypedPattern -> nullptr
  - Type/Expr -> Decl (unless it's `AnyPattern`, in which case -> nullptr)

`Active` would start as `Decl(a)` for the single variable case in order to handle property wrappers. For the tuple case it could start as the container.


I also just noticed that this doesn't currently handle nested `TuplePattern`, ie.
```
let ((a, b), c): ((Int, String), Double) = ((1, ""), 2.0)
```

ileitch replied:
> So there are other Patterns, but we're handling them by creating the mappings in identifyPatternBindingContainerDecls. For the actual mappings themselves we only care about either a single Decl, or a TuplePattern that contains all the NamedPatterns. The TuplePattern has to contain all NamedPatterns as otherwise we would have mapped to Decl (or the current container).

Sorry I don't think I explained myself fully in my last comment. I do understand that for multiple vars they will always be within a `TuplePattern`. However, that point I wanted to make is that using `PointerUnion<void *, Pattern *>` instead allows us to avoid some extra walking in `identifyPatternBindingContainerDecls`.

```
if (auto TP = dyn_cast<TypedPattern>(P)) {
    auto TR = TP->getTypeRepr();

    if (auto TTR = dyn_cast<TupleTypeRepr>(TR)) {
        for (auto n : range(TTR->getNumElements())) {
        auto ET = TTR->getElementType(n);
        setupContainerStackEntryWithVarDecl(Entry, n, ET, MaybeVarDecls);
        }
    } else {
        // Non-tuple type, use all VarDecls in this pattern.
        Entry.DeclMap[TR] = {false, P};
    }
}
```

In the `else` branch here, we can just assign whatever pattern `P` is, it doesn't actually matter. To only use `TuplePattern` in the `PointerUnion` we'd need to check if `P->getSubPattern()` is a `TuplePattern`. It most likely is, however the point is that we don't _need_ to know, simply using `P` is sufficient because all we need is to be able to call `collectVariables` later. Swift is a rapidly evolving language, so my rationale is that the fewer the type checks we need to make, the better; it makes the code less susceptible to breakages when/if new patterns are introduced.

The same applies when handling the initializer:

```
if (auto E = PBD->getInit(i)) {
    if (auto TE = dyn_cast<TupleExpr>(E)) {
        for (auto n : range(TE->getNumElements())) {
        auto E = TE->getElement(n);
        setupContainerStackEntryWithVarDecl(Entry, n, E, MaybeVarDecls);
        }
    } else {
        // Non-tuple initializer, use all VarDecls in this pattern.
        Entry.DeclMap[E] = {false, P};
    }
}
```

That's two places that'd we need to perform some extra walking, when it really isn't necessary. I'm not beholden to using `PointerUnion<void *, Pattern *>`, but if you really think we should constrain it to `TuplePattern`, I'm happy to change it.

> The TypedPattern IMO should always map to nullptr, which would allow us to use the same stack for getParent (ie. a mapping of nullptr == use container, which you currently do for the AnyPattern case).

I'm not sure I follow. For the `PointerUnion` to hold `TypedPattern`, we couldn't then use `TuplePattern`, which is what I think you've been advocating for? Are you also proposing that we merge `EntitiesStack` and `ContainerStack`?

> This also means that ActiveDecls can just be Active and a PointerUnion<Decl *, TuplePattern *> instead.

I did consider making this change, however it'd require that we call `TuplePattern::collectVariables` every time the active container is requested, vs. just once when we walk to the corresponding `TypeRepr`/`Expr` (in `activateContainerDecls`). Given the following var, that would be a lot of extra calls to `collectVariables`:

```
let (a, b): (Int, ((Int, Double, Float, String))) = (intValue, (intValue, doubleValue, floatValue, stringValue))
```

So it boils down to a trade-off between memory usage vs. execution workload. I'm not sure which we should optimize for?

ileitch replied:
> I also just noticed that this doesn't currently handle nested TuplePattern, ie.

This is now handled by the latest commit.

bnbarham replied:
> That's two places that'd we need to perform some extra walking, when it really isn't necessary

Fair enough, `Pattern *` is fine with me.

> So it boils down to a trade-off between memory usage vs. execution workload. I'm not sure which we should optimize for?

`collectVariables` isn't a particularly intensive function, so I'd be happy running it each time.

> This is now handled by the latest commit.

Thanks, I added some comments inline

bnbarham reviewed (commented):
This does seem neater than custom walks, it's just the naming of the balanced methods that's bothering me at the moment. I have no better ideas though 😆

bnbarham commented on lib/Index/Index.cpp:
```diff
@@ -153,6 +154,27 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
   bool isSystemModule = false;
+
+  struct ContainerStackEntry {
+    Decl *const visitedDecl;
```
The rest of the file generally uses UpperCamelCase for members and locals (current LLVM style). I can see a few places that don't, but it would be good to try to remain consistent (ie. use UpperCamelCase).

Also, you could make this a `const Decl *` if you wanted, but I still don't really see the need for the `* const`.

bnbarham commented on lib/Index/Index.cpp:
```diff
@@ -440,6 +464,270 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+  /// are not contained.
+  void identifyContainerVarDecls(const PatternBindingDecl *const bindingDecl,
+                                 ContainerStackEntry &entry) const {
+    if (auto varDecl = bindingDecl->getSingleVar()) {
```
Prefer `auto *` for clarity when it's a pointer. Lots of these

bnbarham commented on lib/Index/Index.cpp:
```diff
@@ -440,6 +464,270 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+      entry.activeDecls = {varDecl};
+    }
+
+    for (const auto patternIndex : range(bindingDecl->getNumPatternEntries())) {
```
No need for const here, it's just a local int.

bnbarham commented on lib/Index/Index.cpp:
```diff
@@ -440,6 +464,270 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+  /// contained by the function. If there is no active container, the references
+  /// are not contained.
+  void identifyContainerVarDecls(const PatternBindingDecl *const bindingDecl,
+                                 ContainerStackEntry &entry) const {
```
`ContainerStackEntry` could be the return for this function (possibly with a rename though). There's a few other functions that are similar as well, ie. `collectPatternVarDecls` could just return the map.

bnbarham commented on lib/Index/Index.cpp:
```diff
@@ -440,6 +464,270 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+        continue;
+      }
+
+      if (auto typedPattern = dyn_cast<TypedPattern>(pattern)) {
```
For the uppercase change - it's pretty common to just use eg. `TP` here

bnbarham commented on lib/Index/Index.cpp:
```diff
@@ -440,6 +464,270 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+    }
+  }
+
+  void collectPatternVarDecls(const Pattern *const pattern,
```
Again, no need for `* const`. There's a bunch more cases like this

bnbarham replied:
Interesting idea with the `path`. It'd be nice to avoid all the string manipulation and copying though.

I haven't thought this through completely, but since we have a guaranteed order here, could we instead construct a vector of {`PointerUnion<Decl *, Pattern*>` and number of children} in the initial collection?

`let (a, ((b, c), d), e)` ->
```
[
  Decl(a),
  {Pattern(bcd), 4},
    {Pattern(bc), 2},
      Decl(b), Decl(c),
    Decl(d),
  Decl(e)
]
```

Then for associating the type and expression, iterate through the vector and skip the number of children if the type/expression isn't a Tuple. So for `: (Int, (SomeAlias, Int), String) = (1, foo(), "")` ->
Walking types:
  - `Int` -> associate to current index and skip children (0 so no skip), go to next index
  - `Tuple` -> expect a pattern (have to decide what to do if not), enter type and go to next index
  - `Tuple` -> expect pattern, enter type and go to next index
  - `Int` -> associate, skip, go to next index
  - `String -> ditto
  - `Int` -> ditto
  - `String` -> ditto

Walking expression:
  - `1` -> associate, skip, go to next index
  - `foo()` -> associate, skip (4 children in this case), go to next index
  - `""` -> associate, skip, go to next index

bnbarham commented on lib/Index/Index.cpp:
```diff
@@ -153,6 +154,27 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+    SmallVector<const Decl *, 6> activeDecls{};
+    llvm::DenseMap<
+        const void *,
+        std::pair<bool, llvm::PointerUnion<const Decl *, const Pattern *>>>
```
Is the `bool` needed here? It seems like a `nullptr` represents the same thing - all `{false, *}` have a value and the only `true` is `{true, nullptr}`

bnbarham commented on test/Index/roles-contained.swift:
```diff
@@ -0,0 +1,277 @@
+// CHECK: [[@LINE-8]]:65 | function/acc-get/Swift | getter:stringValue | {{.*}} | Ref,Call,Impl,RelCont | rel: 1
+// CHECK-NEXT:  RelCont | variable/Swift | tupleIgnoredSiblingElement | {{.*}}
+
+let (tupleNestedElementA, (tupleNestedElementB, (tupleNestedElementC, tupleNestedElementD))): (Int, (String, (Float, Double))) = (intValue, (stringValue, (floatValue, doubleValue)))
```
Would be good to have a nested case where one of the tuples (or multiple) is the return of a function

bnbarham reviewed (commented):
> @bnbarham Sorry for the long pause on this! I started a new job and moved house 😅

No problem! Sounds exciting (and stressful).

Latest change is much simpler, nice :). And thanks for cleaning up all those names!

ileitch commented on lib/Index/Index.cpp:
```diff
@@ -153,6 +154,25 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+    // be associated with the key type. In most cases there will be only one; a
+    // function or single named pattern binding. However in the case of an
+    // expanded tuple pattern binding, there will be multiple.
+    SmallVector<const Decl *, 6> ActiveDecls{};
```
Removing this introduces a warning:
```
Index.cpp:486:34: warning: missing field 'ActiveDecls' initializer [-Wmissing-field-initializers]
      ContainerStackEntry Entry{D};
```

ileitch commented on lib/Index/Index.cpp:
```diff
@@ -565,14 +832,19 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+  void addContainedByRelationIfContained(IndexSymbol &Info) {
+    for (auto D : activeContainerDecls()) {
+      addRelation(Info, (unsigned)SymbolRole::RelationContainedBy,
+                  const_cast<Decl *>(D));
```
That's not possible, since an `AbstractFunctionDecl` can also be a container.

ileitch commented on lib/Index/Index.cpp:
```diff
@@ -153,6 +154,25 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+  typedef llvm::Optional<
+      llvm::PointerUnion<const VarDecl *, const TuplePattern *>>
+      PatternElement;
+  typedef SmallVector<PatternElement, 6> PatternElements;
```
I'm not sure we did. I just picked 6 since it seemed like a reasonable number to cover most TuplePatterns. Happy to tweak if you think more or less would be better.

bnbarham replied:
It'd be good to get some data on this, ie. by looking at tuple patterns in the source compatibility suite. 4 is probably plenty for most cases I'd think (?)

ileitch replied:
Moot point now with the latest changes.

ileitch commented:
@bnbarham Latest change is ready for your perusal. I've switched to on-request collection, and moved everything into a class. Also managed to simplify things still further, we no longer need to keep track of `Pattern` in the map, everything is just a `Decl`.

ileitch commented on lib/Index/Index.cpp:
```diff
@@ -439,6 +461,253 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+    }
+  }
+
+  void associatePatternElement(PatternElement Element, size_t Index,
```
I did also consider using a PointerUnion, but came to the same conclusion. If anything, the unpacking just makes the code a little more verbose.

bnbarham reviewed (commented):
Thanks, this is looking pretty good :). Sorry about the confusion re. StackEntry!

bnbarham commented on lib/Index/Index.cpp:
```diff
@@ -145,6 +146,243 @@ struct IndexedWitness {
+    if (Stack.empty())
+      return {};
+
+    StackEntry Entry = Stack.back();
```
And I missed this one, but should also be `StackEntry &Entry` to avoid the copy.

ileitch replied:
Doing this requires that the `const` qualifier be removed from the function, and in turn a bunch of others. I've no idea which is more preferable.

ileitch commented on lib/Index/Index.cpp:
```diff
@@ -145,6 +146,243 @@ struct IndexedWitness {
+    return MapEntry->second;
+  }
+
+  void activate(ActivationKey K) {
```
Maybe just `activateContainersFor` would be more correct, since this method doesn't _set_ the containers themselves (being extra pedantic here), but it does _activate_ them, by way of setting the `ActiveKey`.

ileitch commented on lib/Index/Index.cpp:
```diff
@@ -145,6 +146,243 @@ struct IndexedWitness {
+      // This pattern consists of a single AnyPattern (i.e let _).
+      associateAnyPattern(K, Entry);
+    } else {
+      ContainerDecls Decls;
+      P->forEachVariable([&](VarDecl *D) { Decls.push_back(D); });
+      Entry.ContainerMap[K] = Decls;
```
No this is my fault. I got carried away with trying to simplify things and in the process managed to revert my previous on-request collection changes 🙃 .

ileitch commented:
Latest changes do on-request collection for real this time 😅. Also removed all of the vectors.

bnbarham reviewed (commented):
Looking good!

Would you be able to compare the performance and index size before/after this change? Mostly just looking for a sanity check with swift-frontend `-typecheck -index-store-path ...` invocations for a reasonably large file/s. Possibly even with `indexLocals()` temporarily returning `true` just to cover more cases. There's some large-ish files in projects in https://github.com/apple/swift-source-compat-suite.

@benlangmuir and @rintaro did you have any other comments here?

benlangmuir commented:
Nothing to add to @bnbarham 's review. Thanks!

ileitch commented:
I used Periphery's [PeripheryKit](https://github.com/peripheryapp/periphery/tree/master/Sources/PeripheryKit) module for this benchmark. [This](https://gist.github.com/ileitch/4696818c437e287635b61942a1c4f5b7) is the script I used.

4327b2a71b50e32 is the latest commit from this branch.
50b44f063e0f745 is the base commit from main.

| Toolchain                      | Store size | Run 1  | Run 2  | Run 3  | Run 4  | Run 5  | Avg.   |
| ------------------------------ | ---------- | ------ | ------ | ------ | ------ | ------ | ------ |
| 4327b2a71b50e32                | 27632      | 16.991 | 16.960 | 16.907 | 17.788 | 17.289 | 17.187 |
| 4327b2a71b50e32 (index locals) | 28232      | 16.550 | 17.214 | 17.072 | 17.277 | 17.189 | 17.060 |
| 50b44f063e0f745                | 27544      | 17.238 | 17.897 | 17.497 | 16.450 | 16.414 | 17.099 |
| 50b44f063e0f745 (index locals) | 28216      | 16.660 | 17.015 | 16.416 | 16.447 | 16.377 | 16.583 |

Key observations:
* Index size increased by 0.31% without locals, and 0.05% with locals. I'd have expected to see the same % increase with local indexing enabled, and ideas as to why this is?
* Fluctuations in runtime make it hard to detect change. I think it's fair to say though that these changes don't introduce any significant degradation. Any ideas what the cause of this fluctuation might be?

Setup:
* 3.1 GHz Quad-Core Intel Core i7, 16 GB 2133 MHz LPDDR3.
* Closed all apps.
* Disconnected internet and Bluetooth.
* Did not observe any thermal throttling.
* Power adapter connected.

bnbarham commented:
> Index size increased by 0.31% without locals, and 0.05% with locals. I'd have expected to see the same % increase with local indexing enabled, and ideas as to why this is?

Something odd going on there. There's only a 8kb (16 * 512 bytes, `du` is weird) increase, if nothing else I'd expect the same at leas the same increase as not having locals at all. Any ideas @benlangmuir?

Ignoring locals though, 44kb increase in a ~14mb store seems pretty reasonable to me and as you mentioned, no significant decrease in performance.

bnbarham commented:
@swift-ci please test

benlangmuir commented:
> Something odd going on there. There's only a 8kb (16 * 512 bytes, du is weird) increase, if nothing else I'd expect the same at leas the same increase as not having locals at all.

Yeah, certainly weird.  Might be worth dumping all that index data and diff'ing it to make sure nothing is missing.  We would expect no references to be lost here, so

* `4327b2a (index locals)` ⊇ `4327b2a`
* `4327b2a (index locals)` ⊇ `50b44f0 (index locals)`
* `50b44f0 (index locals)` ⊇ `50b44f0`

If the dumps look okay then maybe it's just some weird artifact of the storage -- it's been a long time since I looked at the on-disk layout, but could we have fragmentation or something?

bnbarham commented:
> Might be worth dumping all that index data and diff'ing it to make sure nothing is missing

You can use `c-index-test` for this, ie.
```
c-index-test core -print-record store > main.records
```

benlangmuir commented:
> FWIW this isn't a blocker - it's more out of interest at this point.

Agreed, this shouldn't block the PR at this point.

ileitch commented:
OK this is more reassuring. The increase isn't exactly the same, but it's only off by 2474 bytes, which could just be down to a quirk in the formatting of `c-index-store`.

| Toolchain                      | Records size (bytes) | Change          |
| ------------------------------ | -------------------- | --------------- |
| 4327b2a71b50e32                | 24814301             |                 |
| 4327b2a71b50e32 (index locals) | 25027871             | 213570 (+0.86%) |
| 50b44f063e0f745                | 24784926             |                 |
| 50b44f063e0f745 (index locals) | 25000970             | 216044 (+0.87%) |

bnbarham commented:
One change would be different counts, but there should be more for 4327b2a, so that's still a little odd 🤔.

Did you want to rebase before I merge the PR?

ileitch commented:
Rebased 👍

bnbarham commented:
@swift-ci please test

bnbarham commented:
Thanks @ileitch!

bnbarham commented:
@ileitch unfortunately this called compatibility suite failures. Sorry about that, I forgot to run it over your changed before the merge. We've reverted for now in https://github.com/apple/swift/pull/40190.

It looks like it's crashing inside associatePatternElement (https://ci.swift.org/view/Source%20Compatibility/job/swift-main-source-compat-suite-debug/4342). Would you be able to take a look and put up a new PR?

ileitch commented:
I'll take a look, hopefully on the weekend if I've time.
